# Hyperspy as an automatic metadata extractor

With this notebook, I'd like to assess whether or not the Hyperspy library may be implemented to assist the automatic metadata extraction from SEM and TEM research images and data, in order to more efficiently extract metadata from them. The advantage with using this package, is that it uses its own native schema such that the extracted metadata is already structured in a more or less standardized format. While this schema is useless to us, it does allow us to bypass creating a different map for nearly every single instrument as the input to the mapping service will always be in the same format.

### Load the required packages

In [2]:
import hyperspy.api as hs
import json
import os

Load a folder of test images to see how each reacts to the automatic data extraction

In [3]:
def getImages(folder_path):
    tiff_images_list = []
    for filename in os.listdir(folder_path):
        file_path = os.path.join(folder_path, filename)
        if os.path.isfile(file_path) and filename.lower().endswith('.tif'):
            tiff_images_list.append(os.path.join(folder_path, filename))
    return tiff_images_list

testFolder = '/Users/elias/Desktop/MatWerk_Projects/testImages'

images = getImages(testFolder)
images

['/Users/elias/Desktop/MatWerk_Projects/testImages/Na3FePO43 -04-Zeiss IAM-ESS.tif',
 '/Users/elias/Desktop/MatWerk_Projects/testImages/Nozzle Chip RIE KOH10 Zeiss EVO IMT.tif',
 '/Users/elias/Desktop/MatWerk_Projects/testImages/NK_PA07_1-4_KA-W 136980-Zeiss-IAM ESS.tif',
 '/Users/elias/Desktop/MatWerk_Projects/testImages/Au-Gr_06.tif',
 '/Users/elias/Desktop/MatWerk_Projects/testImages/SEM_image_sample_Thermo_Fisher_Helios_G4_PFIB_CXe.tif',
 '/Users/elias/Desktop/MatWerk_Projects/testImages/SEM_image_sample_FEI_Helios_Nanolab600.tif',
 '/Users/elias/Desktop/MatWerk_Projects/testImages/SEM Image 2 - SliceImage - 001.tif']

# Na3FePO43 -04-Zeiss IAM-ESS.tif 

Zeiss Merlin Instrument

In [16]:
f = hs.load(images[0])

In [17]:
Zeiss_Merlin = f.original_metadata.as_dictionary()

del Zeiss_Merlin['ColorMap']
del Zeiss_Merlin['34119']

Zeiss_EVO

{'NewSubfileType': <FILETYPE.UNDEFINED: 0>,
 'ImageWidth': 1024,
 'ImageLength': 768,
 'BitsPerSample': 8,
 'Compression': <COMPRESSION.NONE: 1>,
 'PhotometricInterpretation': <PHOTOMETRIC.PALETTE: 3>,
 'StripOffsets': (103823,),
 'Orientation': <ORIENTATION.TOPLEFT: 1>,
 'SamplesPerPixel': 1,
 'RowsPerStrip': 4294967295,
 'StripByteCounts': (786432,),
 'XResolution': (1, 1),
 'YResolution': (1, 1),
 'ResolutionUnit': <RESUNIT.NONE: 1>,
 'SampleFormat': <SAMPLEFORMAT.UINT: 1>,
 'CZ_SEM': {'': (0,
   0,
   0,
   5.705307e-07,
   195.695,
   6,
   10000.0,
   3.294,
   5e-10,
   0.00953088,
   1,
   5.705307e-07,
   195.695,
   6,
   10000.0,
   3.294,
   5e-10,
   0.00953088,
   2,
   5.705307e-07,
   195.695,
   6,
   10000.0,
   3.294,
   5e-10,
   0.00953088,
   3,
   5.705307e-07,
   195.695,
   6,
   10000.0,
   3.294,
   5e-10,
   0.00953088,
   799),
  'dp_vent_invalid_reason': ('Vent inhibit', 'Beam Present'),
  'dp_optimode': ('OptiBeam Mode', 'Analysis'),
  'dp_fixed_aperture'

We can see simply loading the tiff image and parsing it as a dictionary using built-in hyperspy functions formats it nicely. We can also see that each element of the schema can be found within it, meaning it successfully passes this test. The conversion table for the "variable names" for each metadata variable can be [found here](https://docs.google.com/spreadsheets/d/1f_9qKa2BbA5_q47ild_fZeQFKPUJKcxcbkYkhje0EF0/edit?usp=sharing).

# Nozzle Chip RIE KOH10 Zeiss EVO IMT.tif

Zeiss EVO instrument

In [6]:
f = hs.load(images[1])

In [7]:
Zeiss_EVO = f.original_metadata.as_dictionary()

del Zeiss_EVO['ColorMap']
del Zeiss_EVO['34119']

Zeiss_EVO

{'NewSubfileType': <FILETYPE.UNDEFINED: 0>,
 'ImageWidth': 1024,
 'ImageLength': 768,
 'BitsPerSample': 8,
 'Compression': <COMPRESSION.NONE: 1>,
 'PhotometricInterpretation': <PHOTOMETRIC.PALETTE: 3>,
 'StripOffsets': (103823,),
 'Orientation': <ORIENTATION.TOPLEFT: 1>,
 'SamplesPerPixel': 1,
 'RowsPerStrip': 4294967295,
 'StripByteCounts': (786432,),
 'XResolution': (1, 1),
 'YResolution': (1, 1),
 'ResolutionUnit': <RESUNIT.NONE: 1>,
 'SampleFormat': <SAMPLEFORMAT.UINT: 1>,
 'CZ_SEM': {'': (0,
   0,
   0,
   5.705307e-07,
   195.695,
   6,
   10000.0,
   3.294,
   5e-10,
   0.00953088,
   1,
   5.705307e-07,
   195.695,
   6,
   10000.0,
   3.294,
   5e-10,
   0.00953088,
   2,
   5.705307e-07,
   195.695,
   6,
   10000.0,
   3.294,
   5e-10,
   0.00953088,
   3,
   5.705307e-07,
   195.695,
   6,
   10000.0,
   3.294,
   5e-10,
   0.00953088,
   799),
  'dp_vent_invalid_reason': ('Vent inhibit', 'Beam Present'),
  'dp_optimode': ('OptiBeam Mode', 'Analysis'),
  'dp_fixed_aperture'

The metadata is extracted from this next test image in exactly the same way with the exact same variable names.

# NK_PA07_1-4_KA-W 136980-Zeiss-IAM ESS.tif

Zeiss Supra 55 FE-SEM Instrument

In [8]:
f = hs.load(images[2])
Zeiss_Supra = f.original_metadata.as_dictionary()

del Zeiss_Supra['ColorMap']
del Zeiss_Supra['34119']

Zeiss_Supra

{'NewSubfileType': <FILETYPE.UNDEFINED: 0>,
 'ImageWidth': 2048,
 'ImageLength': 1536,
 'BitsPerSample': 8,
 'Compression': <COMPRESSION.NONE: 1>,
 'PhotometricInterpretation': <PHOTOMETRIC.PALETTE: 3>,
 'StripOffsets': (100950,),
 'Orientation': <ORIENTATION.TOPLEFT: 1>,
 'SamplesPerPixel': 1,
 'RowsPerStrip': 4294967295,
 'StripByteCounts': (3145728,),
 'XResolution': (1, 1),
 'YResolution': (1, 1),
 'ResolutionUnit': <RESUNIT.NONE: 1>,
 'SampleFormat': <SAMPLEFORMAT.UINT: 1>,
 'CZ_SEM': {'': (0,
   0,
   0,
   1.1165e-07,
   1000.0,
   6,
   4000.0,
   2.28,
   2e-07,
   0.008856386,
   1,
   1.1165e-07,
   1000.0,
   6,
   4000.0,
   2.28,
   2e-07,
   0.008856386,
   2,
   1.1165e-07,
   1000.0,
   6,
   4000.0,
   2.28,
   2e-07,
   0.008856386,
   3,
   1.1165e-07,
   1000.0,
   6,
   4000.0,
   2.28,
   2e-07,
   0.008856386,
   778),
  'dp_vent_invalid_reason': ('Vent inhibit', 'Beam Present'),
  'dp_optimode': ('OptiBeam Mode', 'Resolution'),
  'dp_fixed_aperture': ('VP Apert

This image also passes the test, but it's expected as it's the same instrument as the first image.

# Au-Gr_06.tif

Zeiss CNR-IOM instrument

In [9]:
f = hs.load(images[3])
Zeiss_CNRIOM = f.original_metadata.as_dictionary()

del Zeiss_CNRIOM['ColorMap']

Zeiss_CNRIOM

{'NewSubfileType': <FILETYPE.UNDEFINED: 0>,
 'ImageWidth': 1024,
 'ImageLength': 768,
 'BitsPerSample': 8,
 'Compression': <COMPRESSION.NONE: 1>,
 'PhotometricInterpretation': <PHOTOMETRIC.PALETTE: 3>,
 'StripOffsets': (31700,),
 'SamplesPerPixel': 1,
 'RowsPerStrip': 4294967295,
 'StripByteCounts': (786432,),
 'XResolution': (1, 1),
 'YResolution': (1, 1),
 'ResolutionUnit': <RESUNIT.NONE: 1>,
 'CZ_SEM': {'': (0,
   0,
   0,
   9.183126e-10,
   399850.2,
   6,
   5000.0,
   2.25,
   2e-07,
   0.004021607,
   1,
   9.183126e-10,
   399850.2,
   6,
   5000.0,
   2.25,
   2e-07,
   0.004021607,
   2,
   3.696236e-06,
   99.3409,
   6,
   0.0,
   2.25,
   2e-07,
   0.00246,
   3,
   3.696236e-06,
   99.3409,
   6,
   0.0,
   2.25,
   2e-07,
   0.00246,
   724),
  'dp_vent_invalid_reason': ('Vent inhibit', 'Beam Present'),
  'dp_optimode': ('OptiBeam Mode', 'Resolution'),
  'dp_fixed_aperture': ('VP Aperture', False),
  'dp_input_lut_mode': ('Input LUT Mode', 'Transparent'),
  'dp_bsd_auto

This also works as necessary, with the same variable names.

# SEM_image_sample_Thermo_Fisher_Helios_G4_PFIB_CXe.tif

Helios PFIB instrument

In [10]:
f = hs.load(images[4])
Helios_PFIB = f.original_metadata.as_dictionary()

del Helios_PFIB['StripOffsets']
del Helios_PFIB['StripByteCounts']

Helios_PFIB

{'ImageWidth': 1536,
 'ImageLength': 1094,
 'BitsPerSample': (8, 8, 8),
 'Compression': <COMPRESSION.NONE: 1>,
 'PhotometricInterpretation': <PHOTOMETRIC.RGB: 2>,
 'SamplesPerPixel': 3,
 'RowsPerStrip': 1,
 'XResolution': (96, 1),
 'YResolution': (96, 1),
 'PlanarConfiguration': <PLANARCONFIG.CONTIG: 1>,
 'ResolutionUnit': <RESUNIT.INCH: 2>,
 'ExifTag': {'ImageUniqueID': 'C76330991CBF832D40B3A3749A15EB30'},
 'fei_metadata': {'User': {'Date': '03/17/2022',
   'Time': '11:05:00 AM',
   'User': 'user',
   'UserText': 'HeliosPFIB',
   'UserTextUnicode': '480065006C0069006F0073005000460049004200'},
  'System': {'Type': 'DualBeam',
   'Dnumber': 9952707,
   'Software': '14.5.1.432',
   'BuildNr': 432,
   'Source': 'FEG',
   'Column': 'Elstar',
   'FinalLens': 'Elstar',
   'Chamber': 'xT-SDB',
   'Stage': '110 x 110',
   'Pump': 'TMP',
   'ESEM': 'no',
   'Aperture': 'AVA',
   'Scan': 'PIA 3.0',
   'Acq': 'PIA 3.0',
   'EucWD': 0.004,
   'SystemType': 'Helios G4 PFIB CXe',
   'DisplayWidth': 

Here we start getting a little bit of trouble. The output is nicely formatted, in a neat standardized way, but sadly, different from the previous entries. One could make a map for these too, but there also appear to be some components missing (units for some of the values, for one). It could be that some entries have names I don't recognize, and are therefore not missing, but I couldn't find some of the values.

# SEM_image_sample_FEI_Helios_Nanolab600.tif

Helios Nanolab instrument

In [11]:
f = hs.load(images[5])
Helios_Nanolab = f.original_metadata.as_dictionary()

del Helios_Nanolab['StripOffsets']
del Helios_Nanolab['StripByteCounts']

Helios_Nanolab

{'NewSubfileType': <FILETYPE.UNDEFINED: 0>,
 'ImageWidth': 1024,
 'ImageLength': 943,
 'BitsPerSample': (8, 8, 8),
 'Compression': <COMPRESSION.NONE: 1>,
 'PhotometricInterpretation': <PHOTOMETRIC.RGB: 2>,
 'SamplesPerPixel': 3,
 'RowsPerStrip': 1,
 'XResolution': (68, 1),
 'YResolution': (68, 1),
 'PlanarConfiguration': <PLANARCONFIG.CONTIG: 1>,
 'ResolutionUnit': <RESUNIT.CENTIMETER: 3>,
 'fei_metadata': {'User': {'Date': '07/30/2013',
   'Time': '11:11:11 AM',
   'User': 'user',
   'UserText': '',
   'UserTextUnicode': ''},
  'System': {'Type': 'DualBeam',
   'Dnumber': 'D0512',
   'Software': '3.8.9.1943',
   'BuildNr': 1943,
   'Source': 'FEG',
   'Column': 'Elstar',
   'FinalLens': 'Elstar',
   'Chamber': 'xT-SDB',
   'Stage': '6inch',
   'Pump': 'TMP',
   'ESEM': 'no',
   'Aperture': 'AVA',
   'Scan': 'PIA 1.0',
   'Acq': 'PIA 1.0',
   'EucWD': 0.004,
   'SystemType': 'Helios NanoLab',
   'DisplayWidth': 0.32,
   'DisplayHeight': 0.24},
  'Beam': {'HV': 10000,
   'Spot': '',
   

Same story here, except the metadata seems to be even more limited. In any case, it's not the same format as the first few, meaning another map has to be made.

# SEM Image 2 - SliceImage - 001.tif
This is an image from the PP13 SEM/FIB Tomography dataset.

In [12]:
f = hs.load(images[6])
SEMFIB = f.original_metadata.as_dictionary()

del SEMFIB['StripOffsets']
del SEMFIB['StripByteCounts']

SEMFIB

{'ImageWidth': 855,
 'ImageLength': 770,
 'BitsPerSample': 8,
 'Compression': <COMPRESSION.NONE: 1>,
 'PhotometricInterpretation': <PHOTOMETRIC.MINISBLACK: 1>,
 'SamplesPerPixel': 1,
 'RowsPerStrip': 1,
 'PlanarConfiguration': <PLANARCONFIG.CONTIG: 1>,
 'ExifTag': {'ImageUniqueID': 'F44788842094351134387C8FD66D0A9E'},
 'FEI_SFEG': {},
 'FEI_TITAN': '<?xml version="1.0"?>\r\n<Metadata xmlns:nil="http://schemas.fei.com/Metadata/v1/2013/07" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">\r\n  <Core>\r\n    <Guid>1db1caee-750b-44f5-9269-45f0cb417d6b</Guid>\r\n    <UserID>user</UserID>\r\n    <ApplicationSoftware>xT</ApplicationSoftware>\r\n    <ApplicationSoftwareVersion>0</ApplicationSoftwareVersion>\r\n  </Core>\r\n  <Instrument>\r\n    <ControlSoftwareVersion>14.5.1.432</ControlSoftwareVersion>\r\n    <Manufacturer>FEI Company</Manufacturer>\r\n    <InstrumentClass>Helios G4 PFIB CXe</InstrumentClass>\r\n    <InstrumentID>9952707</InstrumentID>\r\n    <ComputerName>HPN105-MPC</Co

Here is the same story. There is some additional metadata about the instrument hidden in a variable in XML format (as a string...), but it is not very useful data, it's missing units, etc. Meaning one has to either refer to some sort of external documentation in order to accurately map this metadata.

# TEM Images Metadata Extraction

As previously discussed, TEM image metadata needs to be extracted from accompanying files, as Hyperspy will only get the metadata from the TIFF image itself, and not the metadata that is "embedded" within it about the actual project. This is presumably by design, and simply because there isn't actually any project metadata embedded into the tiff's coming from TEM instruments. Luckily, the images are indeed usually (always?) accompanied by such files.

In [13]:
temMetadataDir = "/Users/elias/Desktop/MatWerk_Projects/testImages/Data for TEM-Schema"

def getTemMetadata(folder_path):
    mnetadataFileList = []
    for filename in os.listdir(folder_path):
        file_path = os.path.join(folder_path, filename)
        if os.path.isfile(file_path) and filename.lower().endswith('.emd'):
            mnetadataFileList.append(os.path.join(folder_path, filename))
    return mnetadataFileList

temMetadata = getTemMetadata(temMetadataDir)
temMetadata

['/Users/elias/Desktop/MatWerk_Projects/testImages/Data for TEM-Schema/NEW-dimple-polish-pips 20230302 1127 Camera 3800 x Ceta.emd',
 '/Users/elias/Desktop/MatWerk_Projects/testImages/Data for TEM-Schema/NEW-dimple-polish-pips 20230302 1244 STEM 5300 x HAADF.emd',
 '/Users/elias/Desktop/MatWerk_Projects/testImages/Data for TEM-Schema/NEW-dimple-polish-pips 20230302 1143 STEM 7500 x HAADF.emd',
 '/Users/elias/Desktop/MatWerk_Projects/testImages/Data for TEM-Schema/NEW-dimple-polish-pips 20230302 1146 STEM 5300 x HAADF.emd',
 '/Users/elias/Desktop/MatWerk_Projects/testImages/Data for TEM-Schema/LaserCarbon_VS_Themis300-054 Camera 600 mm Ceta 20210421 1623.emd']

In this case, all the images are from the same instrument, so I only extracted the metadata from one to see how it behaves.

In [14]:
TEM = hs.load(temMetadata[0])
TEM.original_metadata.as_dictionary()

{'Core': {'MetadataDefinitionVersion': '7.9',
  'MetadataSchemaVersion': 'v1/2013/07',
  'guid': '00000000000000000000000000000000'},
 'Instrument': {'ControlSoftwareVersion': '2.15.3',
  'Manufacturer': 'FEI Company',
  'InstrumentId': '3900',
  'InstrumentClass': 'Titan',
  'InstrumentModel': 'Themis',
  'ComputerName': 'TITAN52339000'},
 'Acquisition': {'AcquisitionStartDatetime': {'DateTime': '1677752847'},
  'AcquisitionDatetime': {'DateTime': '1677752847'},
  'BeamType': '',
  'SourceType': 'XFEG'},
 'Optics': {'GunLensSetting': '3',
  'ExtractorVoltage': '4100',
  'AccelerationVoltage': '300000',
  'SpotIndex': '4',
  'C1LensIntensity': '-0.27627953886985779',
  'C2LensIntensity': '0.49731478095054626',
  'C3LensIntensity': '0.28566396236419678',
  'ObjectiveLensIntensity': '0.88349461555480957',
  'IntermediateLensIntensity': '-0.016146063804626465',
  'DiffractionLensIntensity': '0.32825303077697754',
  'Projector1LensIntensity': '0.3777472972869873',
  'Projector2LensIntensit

As no map file exists for a TEM Schema (which is still in the works, awaiting on feedback from INT colleagues), it is not yet verifiable if everything required by the schema is extracted. However, it can be immediately noted that the metadata read from the these `.emd` files is much more complete than that from the TIFFs. If such a file format can be provided by all instruments, it's likely that hyperspy reads it all the same, regardless of which instrument it came from. There are also additional metadata export formats beyond simple tiffs which are meant to accompany the research images generated, and therefore can be "linked" to the image in some way. These also provide us with more options and a better way to standardize the metadata extraction process by requiring a metadata file which is automatically generated by most, if not all, instruments (if possible of course). Currently awaiting feedback from the SEM/TEM experts on whether or not these metadata files may be generated along with the TIFFs on all instruments.

# Conclusion

While Hyperspy is not the end-all solution that we were hoping it would be, it does provide a significant boost and efficiency to the metadata extraction. I would indeed recommend that it be implemented, but it needs to be discussed with Reetu and Rossella in how exactly this could be done. We need to also wait on information from INT about whether or not these file formats can be easily provided, and whether this is even an acceptable solution needs to be discussed. 

Hyperspy does pave the way to a "universal" extractor, but it's not the magic tool we were hoping that it would turn out to be. Its extraction algorithm is eons ahead of my manual extraction, but more research is needed to understand exactly how it is completing this extraction under the hood and why some tiffs are treated differently and not extracted under the same schema as others. It appears that it has different structures/schemas for various metadata formats which it receives, but I'm still looking into it. The next steps would be to reformat the codebase of the existing metadata mappings to be more in line with the polished and published tools (i.e. Nicolas' DICOM mapper), and then look into how Hyperspy fits into this.